### PR TITLE
Add system_reset test in nested environment

### PR DESCRIPTION
**Pull request by qcheng-redhat**
ID: 2128942

Signed-off-by: qcheng <qcheng@redhat.com>

### CODE REVIEW

**Comment by qcheng-redhat** [comment]
(1/1) system_reset.Host_RHEL.m9.u1.v0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.1.0.x86_64.io-github-autotest-qemu.nested_system_reset.q35: PASS (133.61 s)

**Comment by qcheng-redhat** [comment]
Hi @PaulYuuu, Can you help review this patch, when you have time. Thanks.

File: qemu/tests/cfg/nested_system_reset.cfg [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+    type = nested_system_reset
+    virt_test_type = qemu
+    no RHEL.7, RHEL.8.0, RHEL.8.1, RHEL.8.2
+    reboot_cmd = "system_reset"
+    qemu_cmd = "/usr/libexec/qemu-kvm &"
+    pgrep_cmd = "pgrep qemu-kvm"
+    HostCpuVendor.amd:
+        cpu_model_flags += ",+svm"
+    HostCpuVendor.intel:
+        cpu_model_flags += ",+vmx"
```
Comment (PaulYuuu) [inline qemu/tests/cfg/nested_system_reset.cfg:12]: Also check `kvm_probe_module_parameters = "nested=1"`?

File: qemu/tests/cfg/nested_system_reset.cfg [hunk 2]
```diff
@@ -0,0 +1,12 @@
+- nested_system_reset:
+    only Linux
+    type = nested_system_reset
+    virt_test_type = qemu
+    no RHEL.7, RHEL.8.0, RHEL.8.1, RHEL.8.2
+    reboot_cmd = "system_reset"
+    qemu_cmd = "/usr/libexec/qemu-kvm &"
```
[the remark this answers is not in the document]
Comment (qcheng-redhat) [reply qemu/tests/cfg/nested_system_reset.cfg]: Updated with -daemonize.

File: qemu/tests/nested_system_reset.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    :param env: Dictionary with test environment
+    """
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login()
+
+    qemu_cmd = params.get("qemu_cmd")
+    pgrep_cmd = params.get("pgrep_cmd")
+    session.cmd(qemu_cmd, ignore_all_errors=True)
```
[the remark this answers is not in the document]
Comment (qcheng-redhat) [reply qemu/tests/nested_system_reset.py:29]: Updated

File: qemu/tests/cfg/nested_system_reset.cfg (hunk 1 above)
Comment (qcheng-redhat) [reply qemu/tests/cfg/nested_system_reset.cfg:12]: I think this is optional. For x86, Intel and AMD are both nested=1 by default. No sure about other archs ?

File: qemu/tests/nested_system_reset.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    :param env: Dictionary with test environment
+    """
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login()
+
+    sm = utils_package.RemotePackageMgr(session, "qemu-kvm")
+    if not (sm.is_installed("qemu-kvm") or sm.install()):
+        test.fail("qemu-kvm is not installed in the guest vm.")
```
Comment (PaulYuuu) [inline qemu/tests/nested_system_reset.py:25]: Can be simplified to:
```suggestion
    if not utils_package.package_install.(session, "qemu-kvm")
        test.fail("qemu-kvm is not installed in the guest vm.")
```

Comment (qcheng-redhat) [reply qemu/tests/nested_system_reset.py]: Updated.

**Review by PaulYuuu — APPROVED** [review]
For code of x86 part: ACK

File: qemu/tests/cfg/nested_system_reset.cfg [hunk 5]
```diff
@@ -0,0 +1,12 @@
+- nested_system_reset:
+    only Linux
+    type = nested_system_reset
+    virt_test_type = qemu
+    no RHEL.7, RHEL.8.0, RHEL.8.1, RHEL.8.2
+    reboot_cmd = "system_reset"
+    qemu_cmd = "/usr/libexec/qemu-kvm -daemonize"
+    pgrep_cmd = "pgrep qemu-kvm"
```
[the remark this answers is not in the document]
Comment (qcheng-redhat) [reply qemu/tests/cfg/nested_system_reset.cfg]: Updated.

**Comment by MiriamDeng** [comment]
As almost all power hosts were broken in the beaker system I have to give the test result a bit late. Once I have the host I will update my part. Thank a lot.

**Review by fbq815 — APPROVED** [review]

**Comment by MiriamDeng** [comment]
Test passed on ppc64le
 (2/2) Host_RHEL.m8.u7.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.7.0.ppc64le.io-github-autotest-qemu.nested_system_reset: STARTED
 (2/2) Host_RHEL.m8.u7.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.7.0.ppc64le.io-github-autotest-qemu.nested_system_reset: PASS (200.98 s)

**Comment by PaulYuuu** [comment]
Okay, I think I have to point out some issues for non-x86 platforms. cc @lijinlijin 
PPC:

1. The command to launch L1 needs `cap-nested-hv=on` in the command line, then L1 has the capability to launch a KVM L2, otherwise, the L2 is a tcg instance.
2. To avoid Spectre and Mitigation, adding `cap-ibs=workaround` is a plus, but we can skip it at this time.

@MiriamDeng, please Launch a L1 with `cap-nested-hv=on` manually and try `/usr/libexec/qemu-kvm -accel kvm` to launch L2.

s390x:

1. By default s390x needs to set `nested=1 hpage=0` when loading the kvm module to enable nested feature, this is required.
2. Using `/usr/libexec/qemu-kvm` is definitely not allowed to start a qemu process, the process will terminate after launching.
3. Even if we don't have a qemu process, `ps aux | grep qemu-kvm` can also return outputs on an idle host
    ```shell
    $ /usr/libexec/qemu-kvm
    VNC server running on ::1:5900
    $ pgrep qemu-kvm
    $ ps aux | grep qemu-kvm
    root       15479  0.0  0.0 215344   760 pts/0    R+   02:06   0:00 grep --color=auto qemu-kvm
    ```
cc @fbq815 

Hello @qcheng-redhat, so please use `pgrep qemu-kvm` for all platforms, and I think `/usr/libexec/qemu-kvm -accel kvm -no-shutdown -daemonize` is the best choice for all platforms as the default command, other specific parameters let's wait for the feedback. And sorry for the patch being stuck for so long.

**Comment by qcheng-redhat** [comment]
Latest result: 

(1/1) Host_RHEL.m9.u2.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.x86_64.io-github-autotest-qemu.boot.q35: PASS (40.89 s)

**Comment by fbq815** [comment]
Test result on s390x:
RHEL.9.1
JOB ID     : 004b398a1c787dbd7036fd20324be41f847b1a8b
JOB LOG    : /root/avocado/job-results/job-2022-10-13T05.09-004b398/job.log
 (1/1) Host_RHEL.m9.u2.nographic.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.1.0.s390x.io-github-autotest-qemu.nested_system_reset.s390-virtio: STARTED
 (1/1) Host_RHEL.m9.u2.nographic.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.1.0.s390x.io-github-autotest-qemu.nested_system_reset.s390-virtio: PASS (40.00 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2022-10-13T05.09-004b398/results.html
JOB TIME   : 41.14 s

RHEL.9.2
JOB ID     : 491bb69e7571a98d02847257c56f1e4056d85207
JOB LOG    : /root/avocado/job-results/job-2022-10-13T05.18-491bb69/job.log
 (1/1) Host_RHEL.m9.u2.nographic.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.s390x.io-github-autotest-qemu.nested_system_reset.s390-virtio: STARTED
 (1/1) Host_RHEL.m9.u2.nographic.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.s390x.io-github-autotest-qemu.nested_system_reset.s390-virtio: PASS (39.11 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2022-10-13T05.18-491bb69/results.html
JOB TIME   : 40.22 s

**Review by fbq815 — APPROVED** [review]

**Comment by MiriamDeng** [comment]
> cap-nested-hv=on Thanks for the reminder, the command should include "cap-nested-hv=on" and it looks like it's not included in this patch on ppc64le, please help to add it, thanks. [stdlog] -sandbox on [stdlog] -machine pseries,memory-backend=mem-machine_mem [stdlog] -nodefaults On ppc, In L1 guest and run L2 guest 1.with "/usr/libexec/qemu-kvm -accel kvm" [root@dhcp16-215-216 ~]# /usr/libexec/qemu-kvm -accel kvm /usr/libexec/qemu-kvm -accel kvm VNC server running on ::1:5901
> 
> 2.with "/usr/libexec/qemu-kvm -accel kvm -no-shutdown -daemonize " [root@dhcp16-215-216 ~]# /usr/libexec/qemu-kvm -accel kvm -no-shutdown -daemonize eusr/libexec/qemu-kvm -accel kvm -no-shutdown -daemonize [ 107.593701] KVMPPC-UVMEM: No support for secure guests VNC server running on ::1:5900

On ppc,
 (2/2) Host_RHEL.m8.u7.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.7.0.ppc64le.io-github-autotest-qemu.nested_system_reset: FAIL: qemu-kvm process started fail in L1 guest. (149.83 s) 

Could you help to check it again ?

**Comment by qcheng-redhat** [comment]
> cap-nested-hv=on Thanks for the reminder, the command should include "cap-nested-hv=on" and it looks like it's not included in this patch on ppc64le, please help to add it, thanks. [stdlog] -sandbox on [stdlog] -machine pseries,memory-backend=mem-machine_mem [stdlog] -nodefaults On ppc, In L1 guest and run L2 guest 1.with "/usr/libexec/qemu-kvm -accel kvm" [root@dhcp16-215-216 ~]# /usr/libexec/qemu-kvm -accel kvm /usr/libexec/qemu-kvm -accel kvm VNC server running on ::1:5901
> 
> 2.with "/usr/libexec/qemu-kvm -accel kvm -no-shutdown -daemonize " [root@dhcp16-215-216 ~]# /usr/libexec/qemu-kvm -accel kvm -no-shutdown -daemonize eusr/libexec/qemu-kvm -accel kvm -no-shutdown -daemonize [ 107.593701] KVMPPC-UVMEM: No support for secure guests VNC server running on ::1:5900

Hi @MiriamDeng, added cap-nested-hv=on for ppc. And please help confirm the qemu command parameters in L1 guest.  Thanks.

**Comment by MiriamDeng** [comment]
Discussed it with qcheng-redhat, and related commands were updated and it passed on ppc side finally, I can grant ack 
On ppc,
[stdlog] 2022-10-16 22:17:24,880 avocado.test DEBUG|     qemu_cmd = /usr/libexec/qemu-kvm -accel kvm -no-shutdown -daemonize
 -machine pseries,cap-nested-hv=on,memory-backend=mem-machine_mem  \
 (1/1) Host_RHEL.m8.u7.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.7.0.ppc64le.io-github-autotest-qemu.nested_system_reset: STARTED
 (1/1) Host_RHEL.m8.u7.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.7.0.ppc64le.io-github-autotest-qemu.nested_system_reset: PASS (148.38 s)

File: qemu/tests/cfg/nested_system_reset.cfg [hunk 6]
```diff
@@ -0,0 +1,14 @@
+- nested_system_reset:
+    only Linux
+    type = nested_system_reset
+    virt_test_type = qemu
+    no RHEL.7, RHEL.8.0, RHEL.8.1, RHEL.8.2
+    reboot_cmd = "system_reset"
+    qemu_cmd = "/usr/libexec/qemu-kvm -accel kvm -no-shutdown -daemonize"
+    ppc64le:
+        machine_type_extra_params += "cap-nested-hv=on"
```
Comment (PaulYuuu) [inline qemu/tests/cfg/nested_system_reset.cfg:9]:
```suggestion
        machine_type_extra_params += ",cap-nested-hv=on"
```

Comment (qcheng-redhat) [reply qemu/tests/cfg/nested_system_reset.cfg]: Yes, updated.

**Comment by fbq815** [comment]
@qcheng-redhat hi qinghua, please add this parameter in this patch for s390x environment setup:
s390x:
        kvm_module_parameters = "nested=1 hpage=0"

**Review by PaulYuuu — APPROVED** [review]
LGTM

**Pull request merged**